File: LeetCode/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {


        stack<char> st;

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') st.push(s[i]);
            else {
                if (st.empty()) return false;
                if (s[i] == ')' && st.top() == '(') st.pop();
                else if (s[i] == ']' && st.top() == '[') st.pop();
                else if (s[i] == '}' && st.top() == '{') st.pop();
                else return false;
            }
        }

        return st.empty();

       
       
    }
};
// @lc code=end

File: LeetCode/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {

        Stack<Character> st = new Stack<>();

        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                st.push(ch);
            } else if (ch == ')' && !st.isEmpty() && st.peek() == '(') {
                st.pop();
            }

            else if (ch == '}' && !st.isEmpty() && st.peek() == '{') {
                st.pop();
            }

            else if (ch == ']' && !st.isEmpty() && st.peek() == '[') {
                st.pop();
            }

            else {
                return false;
            }

        }

        return st.isEmpty();
        //return st.empty();
        






        

        // Stack<Character> stack = new Stack<Character>();

        // for (char c : s.toCharArray()) {
        // if (c == '(' || c == '{' || c == '[')
        // stack.push(c);

        // else if (c == ')' && !stack.isEmpty() && stack.peek() == '(')
        // stack.pop();

        // else if (c == '}' && !stack.isEmpty() && stack.peek() == '{')
        // stack.pop();

        // else if (c == ']' && !stack.isEmpty() && stack.peek() == '[')
        // stack.pop();

        // else
        // return false;
        // }

        // return stack.isEmpty();

    }
}
// @lc code=end
